app/javascript/packages/pwned-passwords/index.js

```diff
@@ -0,0 +1,136 @@
+import EventEmitter from 'node:events';
+import PQueue from 'p-queue';
+import PairingHeap from './pairing-heap.js';
+
+/**
+ * @typedef DownloadOptions
+ *
+ * @prop {string} rangeStart
+ * @prop {string} rangeEnd
+ * @prop {number} concurrency
+ * @prop {number} maxSize
+ */
+
+/**
+ * @typedef HashPair
+ * @prop {string} hash
+ * @prop {number} occurrences
+ */
+
+const API_ROOT = 'https://api.pwnedpasswords.com/range/';
+
+export class Downloader extends EventEmitter {
+  /** @type {string} */
+  rangeStart;
+
+  /** @type {string} */
+  rangeEnd;
+
+  /** @type {PQueue} */
+  downloaders;
+
+  /** @type {number} */
+  maxSize;
+
+  /** @type {PairingHeap<HashPair>} */
+  commonHashes = new PairingHeap((a, b) => a.occurrences - b.occurrences);
+
+  /** @type {TextDecoder} */
+  decoder = new TextDecoder();
+
+  /**
+   * @param {Partial<DownloadOptions>} options
+   */
+  constructor({ rangeStart = '5ba42', rangeEnd = '5bb0a', concurrency = 40, maxSize = 300 }) {
```

Comment:
are these defaults for shorter testing?

Reply:
Yep, eventually I'll swap in the real defaults, once I've got it all working well.

```suggestion
  constructor({ rangeStart = '00000', rangeEnd = 'fffff', concurrency = 40, maxSize = 3_000_000 }) {
```

Reply:
> Yep, eventually I'll swap in the real defaults, once I've got it all working well.

Updated in cb9001f7c.